PocketKit/Sources/PocketKit/Article/Highlights/ArticleComponentHighlights.swift

```diff
@@ -0,0 +1,249 @@
+                tagStack.removeLast()
+            }
+            if character == HighlightConstants.separatorIdentifier {
+                if !tagStack.isEmpty {
+                    var currentComponent = patchedComponents[componentCursor]
+                    var nextComponent = patchedComponents[componentCursor + 1]
+                    tagStack.forEach { _ in
+                        currentComponent.insert(contentsOf: HighlightConstants.highlightEndTag, at: currentComponent.endIndex)
+                        nextComponent.insert(contentsOf: HighlightConstants.highlightStartTag, at: nextComponent.startIndex)
+                    }
+                    patchedComponents[componentCursor] = currentComponent
+                    patchedComponents[componentCursor + 1] = nextComponent
+                }
+                componentCursor += 1
+            }
+            if !scanner.isAtEnd {
+                scanner.currentIndex = patchedBlob.index(after: scanner.currentIndex)
+            }
+        }
+        return patchedComponents
+    }
+}
+
+extension NSAttributedString {
+    func highlighted() -> NSAttributedString {
+        let highlightableString = self.string
+        guard highlightableString.contains(HighlightConstants.commonHighlightTag) else {
+            return self
+        }
+        let scanner = Scanner(string: highlightableString)
+
+        var highlightableRanges = [Range<String.Index>]()
+        var indexStack = [String.Index]()
+
+        while !scanner.isAtEnd {
+            guard scanner.scanUpToString(HighlightConstants.commonHighlightTag) != nil else {
+                continue
+            }
+            let beforeIndex = max(highlightableString.index(before: scanner.currentIndex), highlightableString.startIndex)
+            let character = String(highlightableString[beforeIndex])
+            if character == HighlightConstants.startTagIdentifier {
+                indexStack.append(scanner.currentIndex)
+            }
+            if character == HighlightConstants.endTagIdentifier {
+                let tagIndex = highlightableString.index(before: beforeIndex)
+                if let upperBound = indexStack.popLast() {
+                    let range = Range<String.Index>(uncheckedBounds: (upperBound, tagIndex))
+                    highlightableRanges.append(range)
+                }
+            }
+            if !scanner.isAtEnd {
+                scanner.currentIndex = highlightableString.index(after: scanner.currentIndex)
+            }
+        }
+        let mutable = NSMutableAttributedString(attributedString: self)
+        highlightableRanges.forEach {
+            let nsRange = NSRange($0, in: highlightableString)
+            mutable.addAttribute(.backgroundColor, value: HighlightConstants.highlightColor, range: nsRange)
+            mutable.addAttribute(.foregroundColor, value: HighlightConstants.highlightedTextColor, range: nsRange)
+        }
+        mutable.mutableString.replaceOccurrences(
```

Comment:
Will this affect the highlighted range above or does NSAttributedString know that?

Reply:
NSAttributedString should auto-update its attribute dictionaries for the string replacement.

Reply:
yup, thanks @nzeltzer ! Any changes to the string is reflected to the attributed one. If that wasn't the case, we'd see the tags in the article text